strawberryfields/api/devicespec.py

```diff
@@ -0,0 +1,136 @@
+
+    @property
+    def compiler(self):
+        """list[str]: A list of strings corresponding to Strawberry Fields compilers supported
+        by the hardware device."""
+        return self._compiler
+
+    @property
+    def gate_parameters(self):
+        """dict[str, strawberryfields.circuitspecs.Ranges]: A dictionary of gate parameters
+        and allowed ranges.
+
+        The parameter names correspond to those present in the Blackbird circuit layout.
+
+        **Example**
+
+        >>> spec.gate_parameters
+        {'squeezing_amplitude_0': x=0, x=1, 'phase_0': x=0, 0≤x≤6.283185307179586}
+        """
+        gate_parameters = dict()
+
+        for gate_name, param_ranges in self._gate_parameters.items():
+            # convert gate parameter allowed ranges to Range objects
+            range_list = [[i] if not isinstance(i, Sequence) else i for i in param_ranges]
+            gate_parameters[gate_name] = Ranges(*range_list, variable_name=gate_name)
+
+        return gate_parameters
+
+    def create_program(self, **parameters):
+        """Create a Strawberry Fields program matching the low-level layout of the
+        device.
+
+        Gate arguments should be passed as keyword arguments, with names
+        correspond to those present in the Blackbird circuit layout. Parameters not
+        present will be assumed to have a value of 0.
+
+        **Example**
+
+        Device specifications can be retrieved from the API by using the
+        :class:`~.Connection` class:
+
+        >>> spec.create_program(squeezing_amplitude_0=0.43)
+        <strawberryfields.program.Program at 0x7fd37e27ff50>
+        """
+        bb = blackbird.loads(self.layout)
+
+        # check that all provided parameters are valid
+        for p, v in parameters.items():
+            if p in self.gate_parameters and v not in self.gate_parameters[p]:
+                # parameter is present in the device specifications
+                # but the user has provided a disallowed value
+                raise ValueError(
+                    f"{p} has invalid value {v}. Only {self.gate_parameters[p]} allowed."
+                )
+
+            if p not in self.gate_parameters:
+                raise ValueError(f"Parameter {p} not a valid parameter for this device")
+
+        # determine parameter value if not provided
+        extra_params = set(self.gate_parameters) - set(parameters)
+
+        for p in extra_params:
+            # Set parameter value as the first allowed
+            # value in the gate parameters dictionary.
+            parameters[p] = self.gate_parameters[p].ranges[0].x
+
+        # evaluate the blackbird template
+        bb = bb(**parameters)
+        prog = sf.io.to_program(bb)
+        return prog
+
+    def refresh(self):
+        """Refreshes the device specifications"""
+        device = self._connection._get_device_dict(self.target)
+        self.__init__(target=self.target, connection=self._connection, **device)
```

Comment:
(Minor suggestion): feels a bit strange to call on `__init__` again with certain arguments possibly changing. One way could be factoring out setting layout, modes, compiler and gate_parameters into a separate method and calling that in `__init__` and here too.